DB prevents creating new data permission paths

johnswanson opened the pull request:
We previously had code preventing new data permission paths from being created, but this adds a DB constraint that has the same effect.

Why is this necessary? Imagine we did not have this change in placea, and we have a clustered environment with rolling upgrades. Server A and Server B both start out running v49, and then an upgrade to v50 is triggered.

At the beginning of the upgrade, all data is being read from/written to the Permissions table.

Server A is restarted first, and starts in v50. It begins the migration, copying data permissions from Permissions into DataPermissions.

However, during this process, Server B may accept requests to alter data permissions. Because it's still running v49 code, it will write those changes to the Permissions table and tell the user the change was successful.

But when the rolling upgrade is complete, Server B's changes to the Permissions table will be orphaned - nothing will read them.

With this change, we'll prohibit writing new data permissions paths to the Permissions table at the beginning of the migration process. This way, Server B will respond with errors to any attempt to change data permissions until it's upgraded and the migration is complete.

EDIT: I also ended up deleting code related to "execute" permissions. These permissions were unused. I had test failures related to creating these permissions, and rather than deleting the test and leaving the untested/broken code around, I thought it would be better to just remove it.

johnswanson committed:
DB prevents creating new data permission paths

We previously had code preventing new data permission paths from being
created, but this adds a DB constraint that has the same effect.

Why is this necessary? Imagine we did not have this change in placea,
and we have a clustered environment with rolling upgrades. Server A and
Server B both start out running v49, and then an upgrade to v50 is
triggered.

At the beginning of the upgrade, all data is being read from/written to
the Permissions table.

Server A is restarted first, and starts in v50. It begins the migration,
copying data permissions from Permissions into DataPermissions.

However, during this process, Server B may accept requests to alter data
permissions. Because it's still running v49 code, it will write those
changes to the Permissions table and tell the user the change was
successful.

But when the rolling upgrade is complete, Server B's changes to the
Permissions table will be orphaned - nothing will read them.

With this change, we'll prohibit writing new data permissions paths to
the Permissions table at the beginning of the migration process. This
way, Server B will respond with errors to any attempt to change data
permissions until it's upgraded and the migration is complete.

Each data permissions migration, when rolled back, will place data in
the Permissions table that is invalid according to the constraints added
in a previous migration. That's by design, but it makes rolling back
fail.

Therefore, when we're rolling back any of the permissions migrations,
remove any constraints that exist first. This is idempotent so it's ok
to just try to remove them with each rollback.

mySQL is a special snowflake and doesn't have a built in way to remove a
constraint iff it exists. The simplest way to accomplish this is with a
stored procedure.

Writing a robust, general stored procedure to do this promised to be a
challenge - so for now I just wrote a simple, intentionally non-reusable
stored procedure that always just removes the `no_data_permissions`
constraint (if it exists).

johnswanson commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -5652,7 +5652,61 @@ databaseChangeLog:
+        - sql:
+            dbms: postgresql
+            sql: >-
+              ALTER TABLE permissions ADD CONSTRAINT no_data_permissions CHECK (object NOT LIKE '%/db/%') NOT VALID;
```
"Fun" fact - every database we use has a different way of adding a constraint that _should not_ be checked for validity before it's created (in other words, you can create the constraint and it'll be enforced on inserts/updates but doesn't affect existing rows).

- postgres: `NOT VALID`
- mysql: `NOT ENFORCED`
- h2: `NOCHECK`
- mariadb: no way to do this natively, I used triggers as a workaround

johnswanson committed:
Remove DB execute permissions

These were added but not used anywhere. We'll be deleting them along
with all other data permissions paths in our migrations.

johnswanson commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -5651,8 +5651,93 @@ databaseChangeLog:
+  - changeSet:
+      id: v50.2024-02-27T20:51:15
+      author: johnswanson
+      comment: Create a stored procedure in MySQL to delete a constraint iff it exists
```
MySQL has no built-in way to delete a constraint if and only if it exists. But we need to remove the constraint anytime we are inserting data that may violate it into the database - which means we need to make a way to do so. It turns out that the simplest way is with a stored procedure.

noahmoss commented on resources/migrations/permissions/mariadb_enforce_no_data_permissions_on_update.sql:
```diff
@@ -0,0 +1,11 @@
+CREATE TRIGGER enforce_no_data_permissions_on_update
```
This doesn't hurt to have, of course, but I don't think we ever actually update `permissions` rows. All perms writes are are deletes and then inserts. So this trigger could probably be omitted with no issue.

johnswanson replied:
Good point - I think I will omit this then, as it requires a separate changeset to create and makes things more confusing.

noahmoss reviewed (approved):
Awesome, just a couple comments. Also left a comment on Slack about another potential way to do this on MariaDB without triggers, but if the triggers seem to be working well I'm cool with it

johnswanson commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -5652,7 +5652,151 @@ databaseChangeLog:
 
   - changeSet:
-      id: v50.2024-01-10T03:27:30
+      id: v50.2024-02-20T19:21:04
+      author: camsaul
+      comment: Drop v1 version of v_content view since it references report_card.dataset which we are dropping in next migration
```
I think it's actually an artifact of the diff. I had to add changeSets before the permissions migrations, which meant I had to change the IDs for the permission migration changeSets (so they'd be in order), which meant all the migrations now come after Cam's migration instead of before. The diff thinks I've added Cam's migration up here and deleted it [here](https://github.com/metabase/metabase/pull/39170/files#diff-e6ec4fc8084d80ae2ebedde98c98f358ec7ac526de5c339ebcb454cbda62b221L5758-L5825).

Basically, before things looked like:
- Permission migration (id 1)
- Cam's change (id 2)

I had to change this to:
- Cam's change (id 2)
- My new perms changes (id 3)
- Permission migration (id 4)

It actually occurs to me that before we merge this branch we will need to reorder all these migrations again, to make sure that all the new migrations come after any migrations already in master - e.g. the migration that creates the `data_permissions` table will need to get placed after Cam's changeSet. So these will get all messed up again 😂

noahmoss replied:
Hmm I don't know if migrations necessarily need to go into `master` in order. I've definitely merged migrations in that had other migrations after it...especially when we were reserving migration IDs ahead of time.

johnswanson replied:
I might be wrong, but I thought I'd gotten lint errors before for this, and it looks like the linter at least is intended to block out-of-order migrations.

Do you want me to try without re-arranging, or is this ok as-is?

noahmoss replied:
The IDs in the migration file need to be in order, but they don't need to be merged into master in order, necessarily.

noahmoss replied:
Stepping back a bit — you were saying that you had to add migrations before the perm migrations, which meant that all of them had to come after Cam's migrations.

But could you add the new migrations to have lower IDs than the existing perm migrations? Then you wouldn't have to move them at all. Especially since we're using timestamps now and not reserving IDs.

johnswanson replied:
So the risk I was thinking about here was:
- Cam merges Change 2
- A point release goes out to cloud - it's migrated to 2
- We merge the feature branch with Change 1
- Migration fails because you can't run Change 1 after change 2 has already run.

But this was just my assumption, and apparently... it's wrong! You can run `migrate!` with change 2, then add in change 1 and `migrate!` again and things work just fine. Which is honestly confusing and a little scary to me, because then migrations can run in an unpredictable order - but I guess this is why we try to only include migrations in major releases. 😓 

Thanks for walking me through this until I finally tested my wrong assumptions!

noahmoss replied:
Yeah that's a pretty reasonable assumption to make. I guess as long as all related migrations go in at the same time they'll be run in the same order, but it might be after other migrations with later IDs have already been run. Not sure if we've hit any issues because of that before.

In this case I think we're fine though since all of this should go out in 50, not in any minor releases before then. i.e. we'll merge to `master` but not backport to the 49 release branch.

johnswanson committed:
rm mariadb trigger to enforce data perms on update

This isn't necessary as we never update existing Permissions rows, and
made the migration more complex and confusing. (It's already bad enough!)

johnswanson committed:
Re-reorder migrations
